lib/services/role.go

```diff
@@ -1887,69 +1616,155 @@ func (m *DatabaseNameMatcher) String() string {
 	return fmt.Sprintf("DatabaseNameMatcher(Name=%v)", m.Name)
 }
 
-// CheckAccessToDatabase checks if this role set has access to a particular database.
+// LoginMatcher is a RoleMatcher that checks whether the role's logins
+// match the specified condition.
+type LoginMatcher struct {
+	login *string
+}
+
+func NewLoginMatcher(login string) RoleMatcher {
+	return &LoginMatcher{login: &login}
+}
+
+func (l *LoginMatcher) Match(role types.Role, typ types.RoleConditionType) (bool, error) {
+	logins := role.GetLogins(typ)
+	for _, login := range logins {
+		if *l.login == login {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+// AccessCheckable is the subset of types.Resource required for the RBAC checks.
+type AccessCheckable interface {
+	GetKind() string
+	GetName() string
+	GetMetadata() types.Metadata
+	GetAllLabels() map[string]string
+}
+
+// CheckAccess checks if this role set has access to a particular resource.
+// It attempts to match labels if roleLabelsFn is not nil.
```

Comment:
Expand this comment a bit more, and maybe add an example.
The usage is somewhat clever, especially for people not familiar with method values.
Explain what usually gets passed as `roleLabelsFn` and `matchers` and what the behavior is with no `matchers`.

Alternatively, you can preserve something similar to the old API, but make them all be thin wrappers around `set.checkAccess`. Type-specific wrappers in one place may be slightly less error prone (errors like using the wrong set of labels, forgetting to add a `RoleMatcher` for SSH logins, etc).

Reply:
I saw this comment coming :-) I agree its a bit clever and maybe a bit too easy to make a mistake and thought that strongly typed wrappers would be a good solution.

I do like that this approach doesn't require new methods in the public API every time we add a new resource, though. Teleport has a ton of that and it makes adding new things take longer than it should.

Reply:
Alternative idea - the method could accept a bool for `matchLabels` instead of a func, and if it is set, then we can decide how to fetch the labels internally based on the type and/or kind of `r`. 

Makes it harder to mess up and look at the wrong labels that way.

Reply:
LGTM.